File: Version1/src/main/java/org/example/ApplicationLayer/CompanyService.java
package org.example.ApplicationLayer;

import java.time.LocalDate;
import java.util.List;
import java.util.Optional;
import java.util.Set;
import java.util.UUID;
import java.util.logging.Logger;

import org.example.ApplicationLayer.dto.CompanyDTOs.CompanyAccessResponse;
import org.example.ApplicationLayer.dto.CompanyDTOs.CompanyMembershipResponse;
import org.example.ApplicationLayer.dto.CompanyDTOs.CompanyResponse;
import org.example.ApplicationLayer.dto.CompanyDTOs.HierarchyResponse;
import org.example.ApplicationLayer.dto.CompanyDTOs.InvitationResponse;
import org.example.ApplicationLayer.dto.CompanyDTOs.CompanyPoliciesResponse;
import org.example.ApplicationLayer.dto.CompanyDTOs.SalesReportResponse;
import org.example.ApplicationLayer.dto.CompanyDTOs.EventDtos.DiscountPolicyDto;
import org.example.ApplicationLayer.dto.CompanyDTOs.EventDtos.DiscountRuleDto;
import org.example.ApplicationLayer.dto.CompanyDTOs.EventDtos.PurchasePolicyDto;
import org.example.ApplicationLayer.dto.CompanyDTOs.EventDtos.PurchaseRuleDto;
import org.example.DomainLayer.CompanyAggregate.Company;
import org.example.DomainLayer.CompanyAggregate.CompanyPermission;
import org.example.DomainLayer.DomainException;
import org.example.DomainLayer.NotificationAggregate.INotifier;
import org.example.DomainLayer.PurchaseDomainService;
import org.example.DomainLayer.RolesDomainService;
import org.example.ApplicationLayer.EventService;
import org.example.ApplicationLayer.dto.CompanyDTOs.EventDtos.EventSummaryDto;
import org.example.DomainLayer.PolicyManagment.AgeRule;
import org.example.DomainLayer.PolicyManagment.ConditionalDiscount;
import org.example.DomainLayer.PolicyManagment.CouponCode;
import org.example.DomainLayer.PolicyManagment.DiscountPolicy;
import org.example.DomainLayer.PolicyManagment.IDiscountRule;
import org.example.DomainLayer.PolicyManagment.IPurchaseRule;
import org.example.DomainLayer.PolicyManagment.LoneSeatRule;
import org.example.DomainLayer.PolicyManagment.MaxTicketRule;
import org.example.DomainLayer.PolicyManagment.MinTicketRule;
import org.example.DomainLayer.PolicyManagment.OvertDiscount;
import org.example.DomainLayer.PolicyManagment.PurchaseComposite;
import org.example.DomainLayer.PolicyManagment.PurchasePolicy;
import org.springframework.stereotype.Service;

@Service
public class CompanyService {
    private static final Logger logger = Logger.getLogger(CompanyService.class.getName());

    private final RolesDomainService rolesDomainService;
    private final PurchaseDomainService purchaseDomainService;
    private final INotifier notifier;
    private final EventService eventService;

    public CompanyService(RolesDomainService rolesDomainService,
            PurchaseDomainService purchaseDomainService,
            INotifier notifier,
            EventService eventService) {
        this.rolesDomainService = rolesDomainService;
        this.purchaseDomainService = purchaseDomainService;
        this.notifier = notifier;
        this.eventService = eventService;
    }

    public CompanyResponse createCompany(String founderEmail, String companyName) {
        logger.info("caller=" + founderEmail
                + ", action=createCompany"
                + ", target=RolesDomainService.createCompany"
                + ", params={founderEmail=" + founderEmail + ", companyName=" + companyName + "}");

        if (founderEmail == null || founderEmail.isBlank()) {
            throw new IllegalArgumentException("Founder email is required");
        }
        if (companyName == null || companyName.isBlank()) {
            throw new IllegalArgumentException("Company name is required");
        }

        UUID companyId = rolesDomainService.createCompany(founderEmail, companyName);

        return new CompanyResponse(
                companyId,
                companyName,
                founderEmail,
                0.0,
                true,
                List.of());
    }


    public void closeCompanyAsAdmin(String adminUsername, UUID companyId) {
        logger.info("caller=" + adminUsername
                + ", action=closeCompanyAsAdmin"
                + ", target=RolesDomainService.closeCompanyAsAdmin"
                + ", params={adminUsername=" + adminUsername + ", companyId=" + companyId + "}");

        try {
            if (adminUsername == null || adminUsername.isBlank()) {
                throw new IllegalArgumentException("Admin username is required");
            }
            if (companyId == null) {
                throw new IllegalArgumentException("Company ID is required");
            }

            rolesDomainService.closeCompanyAsAdmin(adminUsername, companyId);

            String owner = rolesDomainService.getCompanyOwner(companyId);

            notifier.notifyUser(owner, "Company: " + companyId + " has been closed");

            logger.info("action=closeCompanyAsAdmin completed successfully"
                    + ", params={adminUsername=" + adminUsername + ", companyId=" + companyId + "}");

        } catch (RuntimeException e) {
            logger.severe("action=closeCompanyAsAdmin failed"
                    + ", caller=" + adminUsername
                    + ", target=RolesDomainService.closeCompanyAsAdmin"
                    + ", params={adminUsername=" + adminUsername + ", companyId=" + companyId + "}"
                    + ", error=" + e.getMessage());
            throw e;
        }
    }


    public InvitationResponse inviteCompanyManager(String ownerUsername,
            UUID companyId,
            String usernameToInvite,
            Set<CompanyPermission> premissions) {
        if (ownerUsername == null || ownerUsername.isBlank()) {
            throw new IllegalArgumentException("Owner username is required");
        }

        UUID invitationId = rolesDomainService.inviteCompanyManager(
                ownerUsername,
                companyId,
                usernameToInvite,
                premissions);

        Company company = rolesDomainService.getCompany(companyId);

        return new InvitationResponse(
                invitationId,
                companyId,
            company.getName(),
                ownerUsername,
                usernameToInvite,
                "MANAGER",
                premissions);
    }

    public InvitationResponse inviteCompanyOwner(String ownerUsername,
            UUID companyId,
            String usernameToInvite) {
        if (ownerUsername == null || ownerUsername.isBlank()) {
            throw new IllegalArgumentException("Owner username is required");
        }

        UUID invitationId = rolesDomainService.inviteCompanyOwner(
                ownerUsername,
                companyId,
                usernameToInvite);

        Company company = rolesDomainService.getCompany(companyId);

        return new InvitationResponse(
                invitationId,
                companyId,
            company.getName(),
                ownerUsername,
                usernameToInvite,
                "OWNER",
                null);
    }

    public void acceptCompanyInvitation(UUID invitationId, String username, UUID companyId) {
        if (invitationId == null) {
            throw new IllegalArgumentException("Invitation ID is required");
        }
        if (username == null || username.isBlank()) {
            throw new IllegalArgumentException("Username is required");
        }
        if (companyId == null) {
            throw new IllegalArgumentException("Company ID is required");
        }

        rolesDomainService.acceptCompanyInvitation(invitationId, username, companyId);
    }

    public void rejectCompanyInvitation(UUID invitationId, String username, UUID companyId) {
        if (invitationId == null) {
            throw new IllegalArgumentException("Invitation ID is required");
        }
        if (username == null || username.isBlank()) {
            throw new IllegalArgumentException("Username is required");
        }
        if (companyId == null) {
            throw new IllegalArgumentException("Company ID is required");
        }

        rolesDomainService.rejectCompanyInvitation(invitationId, username, companyId);
    }

    public List<InvitationResponse> getUserInvitations(String userEmail) {
        if (userEmail == null || userEmail.isBlank()) {
            throw new IllegalArgumentException("User email is required");
        }

        return rolesDomainService.getUserInvitations(userEmail);
    }

    public void removeCompanyMemberAsOwner(String ownerEmail, UUID companyId, String emailToRemove) {
        if (ownerEmail == null || ownerEmail.isBlank()) {
            throw new IllegalArgumentException("Owner email is required");
        }
        if (emailToRemove == null || emailToRemove.isBlank()) {
            throw new IllegalArgumentException("Email to remove is required");
        }
        rolesDomainService.removeCompanyMemberAsOwner(ownerEmail, companyId, emailToRemove);
    }

    public void removeCompanyMemberAsAdmin(String adminUsername, String usernameToRemove) {
        if (adminUsername == null || adminUsername.isBlank()) {
            throw new IllegalArgumentException("Admin username is required");
        }
        if (usernameToRemove == null || usernameToRemove.isBlank()) {
            throw new IllegalArgumentException("Username to remove is required");
        }

        rolesDomainService.removeCompanyMemberAsAdmin(adminUsername, usernameToRemove);
    }

    public void changeManagerPermissions(String ownerUsername,
            UUID companyId,
            String managerUsername,
            Set<CompanyPermission> newPermissions) {
        if (ownerUsername == null || ownerUsername.isBlank()) {
            throw new IllegalArgumentException("Owner username is required");
        }
        if (managerUsername == null || managerUsername.isBlank()) {
            throw new IllegalArgumentException("Manager username is required");
        }

        rolesDomainService.changeManagerPermissions(
                ownerUsername, companyId, managerUsername, newPermissions);

        notifier.notifyUser(managerUsername, "Your manager permissions have changed.");
    }

    public void addPolicyRule(String username,
            UUID companyId,
            Float age,
            Integer minTicket,
            Integer maxTicket,
            Boolean allowLoneSeat) {
        addPolicyRule(
                username,
                companyId,
                Optional.ofNullable(age),
                Optional.ofNullable(minTicket),
                Optional.ofNullable(maxTicket),
                Optional.ofNullable(allowLoneSeat),
                true);
    }

    public void addPolicyRule(String username,
            UUID companyId,
            Optional<Float> age,
            Optional<Integer> minTicket,
            Optional<Integer> maxTicket,
            Optional<Boolean> allowLoneSeat,
            boolean andOr) {
        if (age != null && age.isPresent() && age.get() < 0) {
            throw new IllegalArgumentException("Age must be non negative");
        }
        if (minTicket != null && minTicket.isPresent() && minTicket.get() < 0) {
            throw new IllegalArgumentException("Minimum ticket amount must be non negative");
        }
        if (maxTicket != null && maxTicket.isPresent() && maxTicket.get() < 0) {
            throw new IllegalArgumentException("Maximum ticket amount must be non negative");
        }

        rolesDomainService.addPurchasePolicy(
                username, companyId, age, minTicket, maxTicket, allowLoneSeat, andOr);
    }

    public void deletePolicyRule(String username, UUID companyId, UUID ruleId) {
        if (ruleId == null) {
            throw new IllegalArgumentException("Rule ID is required");
        }

        rolesDomainService.deletePurchasePolicy(username, companyId, ruleId);
    }

    public void addOvertDiscount(String username,
            UUID companyId,
            LocalDate fromDate,
            LocalDate toDate,
            float discountPercent) {
        validateDiscount(toDate, discountPercent);
        rolesDomainService.addOvertDiscount(username, companyId, fromDate, toDate, discountPercent);
    }

    public void addConditionalDiscount(String username,
            UUID companyId,
            LocalDate fromDate,
            LocalDate toDate,
            float discountPercent,
            int requiredTickets,
            int appliedTickets) {
        validateDiscount(toDate, discountPercent);

        if (requiredTickets < 0) {
            throw new IllegalArgumentException("Required tickets must be non negative");
        }
        if (appliedTickets < 0) {
            throw new IllegalArgumentException("Applied tickets must be non negative");
        }

        rolesDomainService.addConditionalDiscount(
                username, companyId, fromDate, toDate,
                discountPercent, requiredTickets, appliedTickets);
    }

    public void addCouponCode(String username,
            UUID companyId,
            LocalDate fromDate,
            LocalDate toDate,
            float discountPercent,
            String code) {
        validateDiscount(toDate, discountPercent);

        if (code == null || code.isBlank()) {
            throw new IllegalArgumentException("Coupon code is required");
        }

        rolesDomainService.addCouponCode(
                username, companyId, fromDate, toDate, discountPercent, code);
    }

    public void removeDiscount(String username, UUID companyId, UUID discountId) {
        if (discountId == null) {
            throw new IllegalArgumentException("Discount ID is required");
        }

        rolesDomainService.removeDiscount(username, companyId, discountId);
    }

    public void rateCompany(UUID userId, UUID companyId, int rating) {
        if (rating < 0 || rating > 5) {
            throw new IllegalArgumentException("Rating must be between 0 and 5");
        }

        try {
            rolesDomainService.rateCompany(userId, companyId, rating);
        } catch (DomainException e) {
            throw new IllegalArgumentException(e.getMessage());
        }
    }

    public HierarchyResponse getCompanyHierarchyMermaid(UUID companyId, String requesterEmail) {
        if (requesterEmail == null || requesterEmail.isBlank()) {
            throw new IllegalArgumentException("Requester email is required");
        }

        String mermaid = rolesDomainService.getCompanyHierarchyMermaid(companyId, requesterEmail);
        return new HierarchyResponse(companyId, mermaid);
    }

    public SalesReportResponse getSalesReportForOwner(String ownerEmail, UUID companyId) {
        if (ownerEmail == null || ownerEmail.isBlank()) {
            throw new IllegalArgumentException("Owner email is required");
        }

        org.example.ApplicationLayer.dto.CompanyDTOs.CompanyAccessResponse access =
            rolesDomainService.getCompanyAccess(companyId, ownerEmail);
        String normalizedRole = access.role == null ? "" : access.role.trim().toLowerCase();
        if (!normalizedRole.equals("owner") && !normalizedRole.equals("founder")) {
            throw new IllegalArgumentException("Only company owners can view the sales report");
        }

        org.example.ApplicationLayer.dto.SalesReport report =
            purchaseDomainService.getSalesReportForOwner(ownerEmail, companyId);
        return new SalesReportResponse(
                companyId,
                ownerEmail,
                report.getEventIds(),
                report.getTicketIds(),
                report.getTotalRevenue());
    }

    public List<org.example.ApplicationLayer.dto.CompanyDTOs.SubordinateEventDto> getEventsManagedBySubordinates(String ownerEmail, UUID companyId) {
        if (ownerEmail == null || ownerEmail.isBlank()) {
            throw new IllegalArgumentException("Owner email is required");
        }

        org.example.ApplicationLayer.dto.CompanyDTOs.CompanyAccessResponse access =
            rolesDomainService.getCompanyAccess(companyId, ownerEmail);
        String normalizedRole = access.role == null ? "" : access.role.trim().toLowerCase();
        if (!normalizedRole.equals("owner") && !normalizedRole.equals("founder")) {
            throw new IllegalArgumentException("Only company owners can view subordinate events");
        }

        List<String> managers = rolesDomainService.getOwnerAndSubordinatesUsernames(companyId, ownerEmail);

        List<org.example.ApplicationLayer.dto.CompanyDTOs.SubordinateEventDto> out = new java.util.ArrayList<>();
        java.util.Set<java.util.UUID> seen = new java.util.HashSet<>();
        for (String mgr : managers) {
            List<EventSummaryDto> events = eventService.getEventsForUserInCompany(mgr, companyId);
            for (EventSummaryDto e : events) {
                if (!seen.contains(e.eventId())) {
                    seen.add(e.eventId());
                    out.add(new org.example.ApplicationLayer.dto.CompanyDTOs.SubordinateEventDto(
                            e.eventId(), e.companyId(), e.companyName(), e.companyRating(),
                            e.name(), e.artist(), e.eventType(), e.date(), e.location(), e.rating(),
                            e.priceMin(), e.priceMax(), e.availableTickets(), e.totalTickets(),
                            mgr
                    ));
                }
            }
        }
        return out;
    }

    public CompanyPoliciesResponse getCompanyPolicies(UUID companyId, String userEmail) {
        if (companyId == null) {
            throw new IllegalArgumentException("Company ID is required");
        }
        if (userEmail == null || userEmail.isBlank()) {
            throw new IllegalArgumentException("Email is required");
        }

        rolesDomainService.getCompanyAccess(companyId, userEmail);
        Company company = rolesDomainService.getCompany(companyId);

        return new CompanyPoliciesResponse(
                company.getId(),
                company.getName(),
                toPurchasePolicyDto(company.getPurchasePolicy()),
                toDiscountPolicyDto(company.getDiscountPolicy()));
    }

    private void validateDiscount(LocalDate toDate, float discountPercent) {
        if (toDate == null) {
            throw new IllegalArgumentException("toDate is required");
        }
        if (toDate.isBefore(LocalDate.now())) {
            throw new IllegalArgumentException("toDate is before today");
        }
        if (discountPercent < 0.0f || discountPercent > 100.0f) {
            throw new IllegalArgumentException("Discount percent must be between 0 and 100");
        }
    }

    public List<CompanyMembershipResponse> getUserCompanies(String userEmail) {
        if (userEmail == null || userEmail.isBlank()) {
            throw new IllegalArgumentException("Email is required");
        }
        logger.info("caller=" + userEmail
                + ", action=getUserCompanies"
                + ", target=RolesDomainService.getUserCompanies"
                + ", params={userEmail=" + userEmail + "}");
        return rolesDomainService.getUserCompanies(userEmail);
    }

    public CompanyAccessResponse getCompanyAccess(UUID companyId, String userEmail) {
        if (companyId == null) {
            throw new IllegalArgumentException("Company ID is required");
        }
        if (userEmail == null || userEmail.isBlank()) {
            throw new IllegalArgumentException("Email is required");
        }

        return rolesDomainService.getCompanyAccess(companyId, userEmail);
    }

    private static PurchasePolicyDto toPurchasePolicyDto(PurchasePolicy purchasePolicy) {
        List<PurchaseRuleDto> out = new java.util.ArrayList<>();
        collectPurchaseLeaves(purchasePolicy == null ? null : purchasePolicy.getRulesView(), out);
        return new PurchasePolicyDto(out);
    }

    private static void collectPurchaseLeaves(IPurchaseRule rule, List<PurchaseRuleDto> out) {
        if (rule == null) {
            return;
        }

        if (rule instanceof PurchaseComposite composite) {
            collectPurchaseLeaves(composite.getLeftRule(), out);
            collectPurchaseLeaves(composite.getRightRule(), out);
            return;
        }

        if (rule instanceof AgeRule age) {
            out.add(new PurchaseRuleDto(age.getId(), "AGE", age.getMinAge(), null, null, null));
        } else if (rule instanceof MinTicketRule min) {
            out.add(new PurchaseRuleDto(min.getId(), "MIN_TICKETS", null, min.getMinTicket(), null, null));
        } else if (rule instanceof MaxTicketRule max) {
            out.add(new PurchaseRuleDto(max.getId(), "MAX_TICKETS", null, null, max.getMaxTicket(), null));
        } else if (rule instanceof LoneSeatRule lone) {
            out.add(new PurchaseRuleDto(lone.getId(), "LONE_SEAT", null, null, null, lone.isAllowLoneSeat()));
        }
    }

    private static DiscountPolicyDto toDiscountPolicyDto(DiscountPolicy discountPolicy) {
        List<DiscountRuleDto> out = new java.util.ArrayList<>();
        if (discountPolicy != null) {
            for (IDiscountRule rule : discountPolicy.getDiscountRules()) {
                if (rule instanceof OvertDiscount overt) {
                    out.add(new DiscountRuleDto(
                            overt.getId(),
                            "OVERT",
                            overt.getFromDate(),
                            overt.getToDate(),
                            overt.getDiscountPercent(),
                            null,
                            null,
                            null));
                } else if (rule instanceof ConditionalDiscount conditional) {
                    out.add(new DiscountRuleDto(
                            conditional.getId(),
                            "CONDITIONAL",
                            conditional.getFromDate(),
                            conditional.getToDate(),
                            conditional.getDiscountPercent(),
                            conditional.getRequiredTickets(),
                            conditional.getAppliedTickets(),
                            null));
                } else if (rule instanceof CouponCode coupon) {
                    out.add(new DiscountRuleDto(
                            coupon.getId(),
                            "COUPON",
                            coupon.getFromDate(),
                            coupon.getToDate(),
                            coupon.getDiscountPercent(),
                            null,
                            null,
                            coupon.getCode()));
                }
            }
        }

        return new DiscountPolicyDto(out);
    }
}

File: frontend/src/services/companyService.ts
import api from "./api";
import type { EventSummary } from "../types/event";

export type CreateCompanyRequest = {
	founderEmail: string;
	companyName: string;
};

export type CompanyMembership = {
	companyId: string;
	companyName: string;
	role: string;
	status: string;
};

export type CompanyResponse = {
	id: string;
	name: string;
	founderEmail: string;
	rating: number;
	isActive: boolean;
	eventIds: string[];
};

export type CompanyPermissionName =
	| "MANAGE_INVENTORY"
	| "CONFIGURE_LAYOUT"
	| "MANAGE_POLICIES"
	| "CUSTOMER_SERVICE"
	| "VIEW_HISTORY"
	| "REPORTS_GENERATION";

export type CompanyAccessResponse = {
	companyId: string;
	companyName: string;
	userEmail: string;
	role: string;
	status: string;
	grantedPermissions: CompanyPermissionName[];
};

export type CompanyPurchasePolicyRuleKind =
	| "AGE"
	| "MIN_TICKETS"
	| "MAX_TICKETS"
	| "LONE_SEAT";

export type CompanyPurchasePolicyRuleResponse = {
	id: string;
	kind: CompanyPurchasePolicyRuleKind;
	minAge: number | null;
	minTickets: number | null;
	maxTickets: number | null;
	allowLoneSeat: boolean | null;
};

export type CompanyPurchasePolicyResponse = {
	rules: CompanyPurchasePolicyRuleResponse[];
};

export type CompanyDiscountRuleKind = "OVERT" | "CONDITIONAL" | "COUPON";

export type CompanyDiscountRuleResponse = {
	id: string;
	kind: CompanyDiscountRuleKind;
	fromDate: string;
	toDate: string;
	percent: number | null;
	requiredTickets: number | null;
	appliedTickets: number | null;
	code: string | null;
};

export type CompanyDiscountPolicyResponse = {
	rules: CompanyDiscountRuleResponse[];
};

export type CompanyPoliciesResponse = {
	companyId: string;
	companyName: string;
	purchasePolicy: CompanyPurchasePolicyResponse;
	discountPolicy: CompanyDiscountPolicyResponse;
};

export type CompanyHierarchyResponse = {
	companyId: string;
	mermaidChart: string;
};

export type CompanyInvitationResponse = {
	invitationId: string;
	companyId: string;
	companyName: string;
	appointerUsername: string;
	appointeeUsername: string;
	invitationType: string;
	permissions: string[] | null;
};

export type InviteManagerRequest = {
	ownerUsername: string;
	usernameToInvite: string;
	permissions: CompanyPermissionName[];
};

export type InviteOwnerRequest = {
	ownerUsername: string;
	usernameToInvite: string;
};

export type DeleteEventRequest = {
	userEmail: string;
	eventManagerEmail: string;
};

export type ChangeManagerPermissionsRequest = {
	ownerUsername: string;
	managerUsername: string;
	newPermissions: CompanyPermissionName[];
};

export type DeletePolicyRuleRequest = {
	username: string;
};

export type RemoveDiscountRequest = {
	username: string;
};

type EventSummaryResponse = {
	eventId: string;
	companyId: string;
	companyName: string;
	companyRating: number;
	name: string;
	artist: string;
	eventType: string;
	date: string;
	location: string;
	rating: number;
	priceMin: number;
	priceMax: number;
	availableTickets: number;
	totalTickets: number;
};

function toEventSummary(response: EventSummaryResponse): EventSummary {
	return {
		id: response.eventId,
		companyId: response.companyId,
		companyName: response.companyName,
		companyRating: response.companyRating,
		name: response.name,
		artist: response.artist,
		type: response.eventType,
		date: response.date,
		location: response.location,
		rating: response.rating,
		priceMin: response.priceMin,
		priceMax: response.priceMax,
		availableTickets: response.availableTickets,
		totalTickets: response.totalTickets,
	};
}

export async function createCompany(
	request: CreateCompanyRequest,
): Promise<CompanyResponse> {
	const response = await api.post("/companies", request);

	return response.data.data as CompanyResponse;
}

export async function getCompanyHierarchy(
	companyId: string,
	requester: string,
): Promise<CompanyHierarchyResponse> {
	const response = await api.get(`/companies/${companyId}/hierarchy`, {
		params: { requester },
	});

	return response.data.data as CompanyHierarchyResponse;
}

export async function getCompanyPermissions(
	companyId: string,
	userEmail: string,
): Promise<CompanyAccessResponse> {
	const response = await api.get(`/companies/${companyId}/permissions`, {
		params: { userEmail },
	});

	return response.data.data as CompanyAccessResponse;
}

export async function getCompanyPolicies(
	companyId: string,
	userEmail: string,
): Promise<CompanyPoliciesResponse> {
	const response = await api.get(`/companies/${companyId}/policies`, {
		params: { userEmail },
	});

	return response.data.data as CompanyPoliciesResponse;
}

export async function getMyCompanies(userEmail: string): Promise<CompanyMembership[]> {
	const response = await api.get("/companies/me/companies", {
		params: { userEmail },
	});

	return response.data.data as CompanyMembership[];
}

export async function getMyInvitations(userEmail: string): Promise<CompanyInvitationResponse[]> {
	const response = await api.get("/companies/me/invitations", {
		params: { userEmail },
	});

	return response.data.data as CompanyInvitationResponse[];
}

export async function inviteCompanyManager(
	companyId: string,
	request: InviteManagerRequest,
): Promise<CompanyInvitationResponse> {
	const response = await api.post(`/companies/${companyId}/managers/invite`, request);

	return response.data.data as CompanyInvitationResponse;
}

export async function inviteCompanyOwner(
	companyId: string,
	request: InviteOwnerRequest,
): Promise<CompanyInvitationResponse> {
	const response = await api.post(`/companies/${companyId}/owners/invite`, request);

	return response.data.data as CompanyInvitationResponse;
}

export async function acceptInvitation(
	companyId: string,
	invitationId: string,
	username: string,
): Promise<void> {
	await api.post(`/companies/${companyId}/invitations/${invitationId}/accept`, null, {
		params: { username },
	});
}

export async function rejectInvitation(
	companyId: string,
	invitationId: string,
	username: string,
): Promise<void> {
	await api.post(`/companies/${companyId}/invitations/${invitationId}/reject`, null, {
		params: { username },
	});
}

export async function getEventsForUserInCompany(
	userEmail: string,
	companyId: string,
): Promise<EventSummary[]> {
	const response = await api.get(`/events/companies/${companyId}/users`, {
		params: { userEmail },
	});

	const rows = (response.data.data ?? []) as EventSummaryResponse[];
	return rows.map(toEventSummary);
}

export async function deleteEvent(
	eventId: string,
	request: DeleteEventRequest,
): Promise<void> {
	await api.delete(`/events/${eventId}`, {
		data: request,
	});
}

export async function changeManagerPermissions(
	companyId: string,
	request: ChangeManagerPermissionsRequest,
): Promise<void> {
	await api.patch(`/companies/${companyId}/managers/permissions`, request);
}

export async function deletePolicyRule(
	companyId: string,
	ruleId: string,
	request: DeletePolicyRuleRequest,
): Promise<void> {
	await api.delete(`/companies/${companyId}/policy/${ruleId}`, {
		data: request,
	});
}

export async function removeDiscount(
	companyId: string,
	discountId: string,
	request: RemoveDiscountRequest,
): Promise<void> {
	await api.delete(`/companies/${companyId}/discounts/${discountId}`, {
		data: request,
}
);
}

export type RemoveMemberOwnerRequest = {
	ownerEmail: string;
	emailToRemove: string;
};

export async function removeCompanyMemberAsOwner(
	companyId: string,
	request: RemoveMemberOwnerRequest,
): Promise<void> {
	await api.delete(`/companies/${companyId}/members/owner`, {
		data: request,
	});
}
